### ISSUE TITLE
Metadata export renders remote_schemas.yaml incorrectly

### ISSUE THREAD

**Issue by creatorrr** [hasura/graphql-engine#7459]
When re-exporting metadata with remote schema permissions, `remote_schemas.yaml` file was rendered incorrectly. The metadata was applied correctly initially, this is only happening on export. In this case, a large enum was rendered with a few newline characters replaced by an empty backslash. In the diff below, `(-)` and `(+)` indicate the changes `metadata export` made. The resulting yaml file then contains invalid SDL and cannot be applied.

Versions tested: `2.0.5`, `2.0.7`

Diff:
```graphql
- name: relations
  permissions:
  - role: developer
    definition:
      schema: "type Query { callRelations(input: String!, relation: Relation!): RelationsResult!\n}\n\ntype
        RelationsResult { result: [String!]!\n}\n\nenum Relation {AtLocation \n  CapableOf
        \n  Causes \n  CausesDesire \n  CreatedBy \n  DefinedAs \n  DesireOf \n  Desires
        \n  HasA \n  HasFirstSubevent \n  HasLastSubevent \n  HasPainCharacter \n
(-)      \n HasPainIntensity \n  HasPrerequisite \n  HasProperty \n  HasSubevent \n
(-)      \n InheritsFrom \n  InstanceOf \n  IsA \n  LocatedNear \n  LocationOfAction
(+)      \ HasPainIntensity \n  HasPrerequisite \n  HasProperty \n  HasSubevent \n
(+)      \ InheritsFrom \n  InstanceOf \n  IsA \n  LocatedNear \n  LocationOfAction
        \n  MadeOf \n  MotivatedByGoal \n  NotCapableOf \n  NotDesires \n  NotHasA
        \n  NotHasProperty \n  NotIsA \n  NotMadeOf \n  PartOf \n  ReceivesAction
        \n  RelatedTo \n  SymbolOf \n  UsedFor \n}"
```

**Comment by robx** [comment]
I've confirmed this now in the following scenario (will update as I progress):
- running current development graphql-engine, create a remote schema based on `server/tests-py/remote_schemas/nodejs/index.js` modified to use the long enum from this bug report
- export the metadata using hasura cli v2.0.7

**Comment by robx** [comment]
So, I dug into this a bit.

- the issue lies with hasura cli's metadata export (specifically the formatting at https://github.com/hasura/graphql-engine/blob/a835d95b17e540c226365b12ffb8005bd8e347cd/cli/internal/metadataobject/remoteschemas/remote_schemas.go#L157 via gopkg.in/yaml.v2, which introduces weird linewrapping
- this isn't a newline being replaced by a single backspace; instead, it's the yaml package uselessly escaping a space character
- the generated output is messy, but not technically incorrect

I'd argue the behavior in the yaml package might be considered a bug, but... Back to cli folks to decide on what/whether they want to do anything about it.

Here's a snippet to show the yaml encoder's behavior:
```
$ cat x.go
package main

import (
	"gopkg.in/yaml.v2"
	"strings"
)

func main() {
	long := strings.Repeat("x", 100) + "  z \n"
	var v = map[string]string{
		"a": long,
	}
	data, _ := yaml.Marshal(v)
	print(string(data))
}
$ go run x.go
a: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  \ z \n"
```

(Easiest workaround might be to disable linewrapping by calling `yaml.FutureLineWrap()`, but that might affect other parts negatively.)

**Comment by purush7** [comment]
Thanks, @robx for pointing out the cause of the error and `yaml.FutureLineWrap()`

**Comment by purush7** [comment]
Hey @robx, I have found these escape characters are inserted only in case of if the first character after the line wrapped and the next one is also a space.

```
> cat main.go                                                                                                                                                                  
package main

import (
	"strings"

	"gopkg.in/yaml.v2"
)

func main() {
	long := strings.Repeat("x", 90) + "  z \n a    b \n \n\n" + strings.Repeat("x", 90) + "y x \n"
	long2 := strings.Repeat("x", 90) + "  z \n a    b \n \n\n" + strings.Repeat("x", 90) + "y  x \n"
	var v = map[string]string{
		"a": long,
		"b": long2,
	}
	data, _ := yaml.Marshal(v)
	print(string(data))
}


>  go run main.go                                                                                                                                                              
a: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  \ z \n a    b \n \n\nxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxy
  x \n"
b: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  \ z \n a    b \n \n\nxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxy
  \ x \n"
```

and `yaml.FutureLineWrap()` solves the bug

**Comment by mfcochauxlaberge** [comment]
![image](https://user-images.githubusercontent.com/534861/146796577-af178226-a14c-47c0-a908-d04c475371d5.png)
Three YAML libraries in one file? :wink: 

@robx mentions a possible bug from `gopkg.in/yaml.v2`. Maybe upgrading to `gopkg.in/yaml.v3` could help? I'm not familiar with building Hasura, but I'll see what I can do to fix this.

This bug has been slowing us down at lot and we thought it was going to be fixed for Hasura v2.1.0, but it's not the case. :/

**Comment by purush7** [comment]
@mfcochauxlaberge We are using 
- `gopkg.in/yaml.v2` for mainly `MapSlice` and `MapItem` which is absent in `v3`. It is used to preserve the order
- `gopkg.in/yaml.v3` for  Node type
- `github.com/goccy/go-yaml` for  converting YAML byte to JSON byte

PR has been created to address this issue and also the issue of YAML not treating the graphql schema as a multi-line string.
It's in reviewing state. You might expect the change in the next release
